File: node/shared/kafka/producer_.ts
import { Service as Microservice } from '../../shared/types/service';
import Kafka, { Producer as RdKafkaProducer, LibrdKafkaError, ProducerGlobalConfig, TopicConfig } from 'node-rdkafka';

interface CustomMessageFormat {
  a: string;
}

class Producer implements Microservice.Producer {

  public name: string;
  private producer: RdKafkaProducer;
  private topic: any;

  constructor(options?: any) {
    this.producer = this.create(options);
    this.topic = options.topic;
    this.name = options.topic.topics[0];
  }

  public async start(): Promise<void> {
    try {
      await new Promise<void>((resolve, reject) => {
        this.producer.connect({}, (err: LibrdKafkaError) => {
          if (err) {
            reject(err);
          } else {
            resolve();
          }
        });
      });
    } catch (error) {
      console.error(`Error connecting ${this.name} producer:`, error);
    }
  }

  public async shutdown(): Promise<void> {
    return new Promise<void>((resolve, reject) => {
      this.producer.disconnect((err: Error) => {
        if (err) {
          console.error(`Error disconnecting ${this.name} producer:`, err);
          reject(err);
        } else {
          console.log(`${this.name} producer disconnected.`);
          resolve();
        }
      });
    });
  }

  public async send(message: CustomMessageFormat): Promise<void> {
    return new Promise<void>((resolve, reject) => {
      this.producer.produce(
        this.topic,
        null,
        Buffer.from(JSON.stringify(message)),
        null,
        Date.now(),
        (err: LibrdKafkaError, offset: any) => {
          if (err) {
            console.error(`Error sending message from ${this.name} producer:`, err);
            reject(err);
          } else {
            console.log(`Message sent to ${this.name} at offset ${offset}`);
            resolve();
          }
        }
      );
    });
  }

  // Additional method for sending batches
  public async sendBatch(messages: Array<CustomMessageFormat>): Promise<void> {
    // Implement batch sending logic here
    // Note: The node-rdkafka library doesn't have a direct equivalent for batch sending as in kafkajs
  }

  private create(options: any): RdKafkaProducer {
    const { brokers, clientId, topic } = options;

    const producerOptions: ProducerGlobalConfig = {
      "client.id": clientId,
      "metadata.broker.list": brokers,
      'message.send.max.retries': 10,
      'socket.keepalive.enable': true,
      'queue.buffering.max.messages': 100000,
      'queue.buffering.max.ms': 1000,
      'batch.num.messages': 1000000,
    };

    const producer = new Kafka.Producer(producerOptions);

    // set event listeners
    producer
      .on('ready', () => {
        console.log(`${this.name} producer connected.`);
      })
      .on('event.error', (err: LibrdKafkaError) => {
        console.error(`Error in ${this.name} producer:`, err);
      });

    producer.setPollInterval(100);

    return producer;
  }
}

export default Producer;

File: go/shared/kafka/producer.go
package kafka

import (
	"encoding/json"
	"fmt"
	"log"
	"strings"

	"github.com/NCSU-Microservice-Benchmarking/vision_middleware/go/shared/types"
	"github.com/confluentinc/confluent-kafka-go/kafka"
)

type Producer struct {
	Name     string
	Producer *kafka.Producer
	Topic    string
}

func NewProducer(name string, options types.KafkaOptions) (*Producer, error) {

	config := &kafka.ConfigMap{
		"bootstrap.servers":        strings.Join(options.Brokers, ","),
		"message.send.max.retries": 10,
		"socket.keepalive.enable":  true,
	}

	p, err := kafka.NewProducer(config)
	if err != nil {
		return nil, err
	}
	// Initialize event handlers
	go handleEvents(p)

	return &Producer{
		Name:     name,
		Producer: p,
		Topic:    options.Topics.Producer,
	}, nil
}

func (p *Producer) Start() error {
	fmt.Printf("%s producer connected.", p.Name)
	return nil // No action needed to start in Go
}

func (p *Producer) Send(message types.Request) error {
	deliveryChan := make(chan kafka.Event)

	// Marshal the message to JSON
	newMessage, err := json.Marshal(message)
	if err != nil {
		return err
	}

	// send the message
	err = p.Producer.Produce(&kafka.Message{
		TopicPartition: kafka.TopicPartition{Topic: &p.Topic, Partition: kafka.PartitionAny},
		Value:          newMessage,
	}, deliveryChan)

	if err != nil {
		return err
	}

	// Handle delivery report
	ev := <-deliveryChan
	m := ev.(*kafka.Message)
	if m.TopicPartition.Error != nil {
		return m.TopicPartition.Error
	}

	log.Printf("Message sent to topic %s at offset %v\n", *m.TopicPartition.Topic, m.TopicPartition.Offset)
	return nil
}

// handleEvents handles events from the Kafka producer
func handleEvents(p *kafka.Producer) {
	for e := range p.Events() {
		switch ev := e.(type) {
		case *kafka.Message:
			if ev.TopicPartition.Error != nil {
				log.Printf("Delivery failed: %v\n", ev.TopicPartition.Error)
			} else {
				log.Printf("Delivered message to %s [%d] at offset %v\n",
					*ev.TopicPartition.Topic, ev.TopicPartition.Partition, ev.TopicPartition.Offset)
			}
		}
	}
}

func (p *Producer) Shutdown() error {
	defer p.Producer.Close()
	return nil
}
